### PR TITLE
kie-issues#888: Decision Services & multiple DRDs: Adding a Decision Service to a DRD

### LINKED ISSUE #888: Decision Services & multiple DRDs: Adding a Decision Service to a DRD.
On the DMN Editor, when a Decision Service is added to a DRD, all of its contained Decisions need to be moved inside it or added to the DRD, so that the internal depiction of that Decision Service is exactly the same as all other DRDs, or automatically layout'ed if none is found. However, if one of its contained Decisions it also contained by other Decision Services present on that DRD, and one of them is in expanded form, the Decision Service is added as collapsed, and no Decisions are added or moved.

EDIT: When the collapsed case happens, we also need to check for any Decisions that are contained by the Decision Service that is being added as collapsed. Those Decisions need to be removed from the DRD, as they're "swallowed" by the Decision Service.

### PR DESCRIPTION
**Pull request by tiagobento**
Closes https://github.com/apache/incubator-kie-issues/issues/888

Before:

https://github.com/apache/incubator-kie-tools/assets/1584568/8a83a7e7-b2b3-4480-83be-c05f0a418c71

After:

https://github.com/apache/incubator-kie-tools/assets/1584568/d131d864-34e2-41c4-984e-510bdcbf5961

### CODE REVIEW

File: packages/dmn-editor/src/diagram/Diagram.tsx [hunk 1]
```diff
@@ -357,39 +367,61 @@ export const Diagram = React.forwardRef<DiagramRef, { container: React.RefObject
...
+                decisionServiceNamespace: externalNodeDmn.model.definitions["@_namespace"],
+                drdIndex: state.diagram.drdIndex,
+                dropPoint,
+                externalDmnsIndex,
+                thisDmnsDefinitions: state.dmn.model.definitions,
+                thisDmnsIndexedDrd: state.computed(state).indexedDrd(),
+                thisDmnsNamespace: state.dmn.model.definitions["@_namespace"],
+              });
+            } else if (externalDrgElement.__$$element === "decision") {
+              // TODO: Tiago --> Add logic for Decision contained by DS.
```
Comment (tiagobento) [inline packages/dmn-editor/src/diagram/Diagram.tsx:391]: Will be addressed on https://github.com/apache/incubator-kie-issues/issues/894

File: packages/dmn-editor/src/diagram/Diagram.tsx [hunk 2]
```diff
@@ -398,31 +430,65 @@ export const Diagram = React.forwardRef<DiagramRef, { container: React.RefObject
...
+                drdIndex: state.diagram.drdIndex,
+                dropPoint,
+                externalDmnsIndex: state.computed(state).getExternalModelTypesByNamespace(externalModelsByNamespace)
+                  .dmns,
+                thisDmnsDefinitions: state.dmn.model.definitions,
+                thisDmnsIndexedDrd: state.computed(state).indexedDrd(),
+                thisDmnsNamespace: state.dmn.model.definitions["@_namespace"],
+              });
+            } else if (drgElement.__$$element === "decision") {
+              // TODO: Tiago --> Add logic for Decision contained by DS.
```
Comment (tiagobento) [inline packages/dmn-editor/src/diagram/Diagram.tsx:457]: Will be addressed on https://github.com/apache/incubator-kie-issues/issues/894

File: packages/dmn-editor/src/mutations/addExistingDecisionServiceToDrd.ts [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+    if (hasCompleteExpandedDepictionOfDecisionService) {
+      indexedDrd = _indexedDrd;
+      break; // Found a DRD with a complete expanded depiction of the Decision Service.
+    }
+  }
+
+  if (!indexedDrd) {
+    // There's no DRD which inclues a complete expanded depiction of the Decision Service. Let's proceed with auto-layout.
+    // TODO: Tiago
```
Comment (tiagobento) [inline packages/dmn-editor/src/mutations/addExistingDecisionServiceToDrd.ts:133]: Will be addressed on https://github.com/apache/incubator-kie-issues/issues/972

File: packages/dmn-editor/src/store/computed/computeDecisionServiceHrefsByDecisionHrefs.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  thisDmnsNamespace,
+  drgElementsNamespace,
+  drgElements,
+}: {
+  thisDmnsNamespace: string;
+  drgElementsNamespace: string;
+  drgElements: State["dmn"]["model"]["definitions"]["drgElement"];
+}) {
+  drgElements ??= [];
+  const ret = new Map<string, string[]>();
```
Comment (ljmotta) [inline packages/dmn-editor/src/store/computed/computeDecisionServiceHrefsByDecisionHrefs.ts]: Could you please add a comment or a better name?

File: packages/dmn-editor/src/mutations/addExistingDecisionServiceToDrd.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+      !(dsShape["@_isCollapsed"] ?? false) &&
+      containedDecisionHrefsRelativeToThisDmn.every((dHref) => _indexedDrd.dmnShapesByHref.has(dHref));
+
+    if (hasCompleteExpandedDepictionOfDecisionService) {
+      indexedDrd = _indexedDrd;
+      break; // Found a DRD with a complete expanded depiction of the Decision Service.
+    }
+  }
+
+  if (!indexedDrd) {
```
Comment (ljmotta) [inline packages/dmn-editor/src/mutations/addExistingDecisionServiceToDrd.ts:131]: All the logic below shouldn't be inside the for? What will happen if we have multiple DRDs with the decision service on its expaded form?

**Review by ljmotta — COMMENTED** [review]
@tiagobento Great addition! I've made just a few comments inline.

File: packages/dmn-editor/src/store/computed/computeDecisionServiceHrefsByDecisionHrefs.ts (hunk 4 above)
Comment (tiagobento) [reply packages/dmn-editor/src/store/computed/computeDecisionServiceHrefsByDecisionHrefs.ts]: Renamed

File: packages/dmn-editor/src/mutations/addExistingDecisionServiceToDrd.ts (hunk 5 above)
Comment (tiagobento) [reply packages/dmn-editor/src/mutations/addExistingDecisionServiceToDrd.ts:131]: We only use the first DRD with the complete, expanded depiction of the DS as reference. The for above this line will break once one is found.

**Review by ljmotta — APPROVED** [review]

**Pull request merged**